witness/witness/cmd/feeder/main.go

```diff
@@ -0,0 +1,177 @@
+	logID string
+	wsth  *ct.SignedTreeHead
+	c     *jsonclient.JSONClient
+	w     wh.Witness
+}
+
+func main() {
+	flag.Parse()
+	ctx := context.Background()
+	// Set up the witness client.	
+	if *logPK == "" {
+		glog.Exit("--log_pk must not be empty")
+	}
+	var w wh.Witness
+	pemPK, err := ioutil.ReadFile(*logPK)
+	if err != nil {
+		glog.Exitf("Failed to read public key from file: %v", err)
+	}
+	pk, _, _, err := ct.PublicKeyFromPEM(pemPK)
+	if err != nil {
+		glog.Exitf("Failed to create public key: %v", err)
+	}
+	sv, err := ct.NewSignatureVerifier(pk)
+	if err != nil {
+		glog.Exitf("Failed to create signature verifier: %v", err)
+	}
+	if wURL, err := url.Parse(*witness); err != nil {
+		glog.Exitf("Failed to parse witness URL: %v", err)
+	} else {
+		w = wh.Witness{
+			URL:      wURL,
+			Verifier: *sv,
+		}
+	}
+	// Now set up the log client.
+	if *logID == "" {
+		glog.Exit("--log_id must not be empty")
+	}
+	if *log == "" {
+		glog.Exit("--log must not be empty")
+	}
+	opts := jsonclient.Options{PublicKey: string(pemPK)}
+	c, err := jsonclient.New(*log, http.DefaultClient, opts)
+	if err != nil {
+		glog.Exitf("Failed to create JSON client: %v", err)
+	}
+	// Create the feeder with no initial witness STH.
+	feeder := feeder{
+		logID: *logID,
+		c:     c,
+		w:     w,
+	}
+
+	tik := time.NewTicker(*pollInterval)
+	for {
+		var wSize uint64
+		if feeder.wsth != nil {
+			wSize = feeder.wsth.TreeSize
+		} else {
+			wSize = 0
+		}
+		glog.V(2).Infof("Tick: start feedOnce (witness size %d)", wSize)
+		if err := feeder.feedOnce(ctx); err != nil {
+			glog.Warningf("Failed to feed: %v", err)
+		}
+		glog.V(2).Infof("Tick: feedOnce complete (witness size %d)", wSize)
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-tik.C:
+		}
+	}
+}
+
+// feedOnce attempts to update the STH held by the witness to the latest STH
+// provided by the log.
+func (f *feeder) feedOnce(ctx context.Context) error {
+	// Get and parse the latest STH from the log.
+	var sthResp ct.GetSTHResponse
+	_, csthRaw, err := f.c.GetAndParse(ctx, ct.GetSTHPath, nil, &sthResp)
+	if err != nil {
+		return fmt.Errorf("failed to get latest STH: %v", err)
+	}
+	csth, err := sthResp.ToSignedTreeHead()
+	if err != nil {
+		return fmt.Errorf("failed to parse response as STH: %v", err)
+	}
+	var wSize uint64
+	if f.wsth != nil {
+		wSize = f.wsth.TreeSize
+	} else {
+		wSize = 0
+	}
+	if wSize >= csth.TreeSize {
+		glog.V(1).Infof("Witness size %d >= log size %d - nothing to do", wSize, csth.TreeSize)
+		return nil
+	}
+
+	glog.Infof("Updating witness from size %d to %d", wSize, csth.TreeSize)
+	// If we want to update the witness then let's get a consistency proof.
+	params := map[string]string{
+		"first":  strconv.FormatUint(wSize, 10),
+		"second": strconv.FormatUint(csth.TreeSize, 10),
+	}
+	var pfResp ct.GetSTHConsistencyResponse
+	var pf [][]byte
+	if wSize > 0 {
+		if _, _, err := f.c.GetAndParse(ctx, ct.GetSTHConsistencyPath, params, &pfResp); err != nil {
```

Comment:
Could you use the [`logclient.GetSTHConsistency`](https://github.com/google/certificate-transparency-go/blob/master/client/logclient.go) func or do you need to use the `jsonclient` directly?

Reply:
Oh cool, I didn't know you could call those underlying functions directly!  I've now moved to using `logclient` and am calling the higher-level `GetSTHConsistency` for the consistency proof and the lower-level `GetAndParse` for the STH.